a2i2/surround
# FEATURE: Cloud Compute with Google Cloud

Integrates Google Cloud AI Platform with the generated project, allowing the running of batch/training jobs in the cloud.

**TO TEST THIS PR**:
- Create a Google Cloud Platform account and project
- Enable the AI Platform Training & Prediction API, Google Cloud Storage JSON API, and Google Container Registry API.
- Create a service account for the project and download the JSON credentials file
- Ensure the service account has IAM permissions called `ML Engine Developer` & `Storage Admin`
- Set the environment var `GOOGLE_APPLICATION_CREDENTIALS` to the path to the JSON file
- Remove `surround==...` from both `requirements.txt` and `requirements-gpu.txt` in the generated project.
- In `Dockerfile` and `Dockerfile-gpu` add the following:
    ```
    # Install the experimental version of Surround
    RUN git clone https://github.com/a2i2/surround.git && \
        cd surround && \
        git fetch origin pull/217/head:experiment-platform && \
        git checkout experiment-platform && \
        pip3 install --upgrade -r requirements.txt && \
        export VERSION_TAG=0.0.13 && \
        python3 setup.py install && \
        cd .. && rm -r surround
    ```
- Then build images and deploy! (see below)

Build and push Docker images that are used in the cloud (required to run cloud jobs):
```
$ surround run buildCompute
$ surround run buildComputeGPU
```

Run a cloud job (will throw an error if no images have been built):
```
$ surround run batchCompute
$ surround run trainCompute

$ surround run batchComputeGPU
$ surround run trainComputeGPU
```

Manage/list cloud jobs while they're running:
```
$ surround run killCompute <job_id>
$ surround run statusCompute <job_id>
$ surround run listCompute
```

New files and their purpose:
- `Dockerfile-compute` - Dockerfile for CPU compute image
- `Dockerfile-compute-gpu` Dockerfile for GPU compute image (base: `nvidia/cuda:1.0-cudnn7`)
- `requirements-gpu.txt` - Requirements file used when building GPU image

New default configuration properties (in config.yaml):
```
compute:
  cpu:
    scale-tier: BASIC
    region: us-east1
  gpu:
    scale-tier: BASIC_GPU
    region: us-east1
```
